lib/galaxy/managers/metrics.py

```diff
@@ -0,0 +1,159 @@
+
+    def __init__(self, app: StructuredApp) -> None:
+        self._app = app
+        #: set to true to send additional debugging info to the log
+        self.debugging = True
+
+    def create(self, trans, payload: CreateMetricsPayload):
+        """
+        Record any metrics sent and return some status object.
+
+        .. note:: Anonymous users can post metrics
+
+        :type   payload: dict
+        :param  payload: (optional) dictionary structure containing:
+            * metrics:          a list containing dictionaries of the form
+
+                namespace:       label indicating the source of the metric
+                time:            isoformat datetime when the metric was recorded
+                level:           an integer representing the metric's log level
+                args:            a json string containing an array of extra data
+
+        :rtype:     dict
+        :returns:   status object
+        """
+        user_id = trans.user.id if trans.user else None
+        session_id = trans.galaxy_session.id if trans.galaxy_session else None
+        parsed_gen = self._parse_metrics(payload.metrics, user_id, session_id)
+        self._send_metrics(trans, parsed_gen)
+        response = self._get_server_pong(trans)
+        return response
+
+    def _parse_metrics(
+        self,
+        metrics: Optional[List[Metric]] = None,
+        user_id=None,
+        session_id=None
+    ) -> TimeSeriesTupleGenerator:
+        """
+        Return a generator yielding the each given metric as a tuple:
+            * label:    the namespace of the metric
+            * time:     datetime of the metric's creation
+            * kwargs:   a dictionary containing:
+                ** level:   the log level of the metric
+                ** user:    the user associated with the metric
+                            (will be None if anonymous user)
+                ** session: the session of the current user
+        """
+        metrics = metrics or []
+        for metric in metrics:
+            label = metric.namespace
+            time = self._deserialize_isoformat_date(metric.time)
+            kwargs = {
+                'level': metric.level,
+                'args': metric.args,
+                'user': user_id,
+                'session': session_id
+            }
+            yield (label, time, kwargs)
+
+    def _send_metrics(self, trans, metrics: TimeSeriesTupleGenerator) -> None:
+        """
+        Send metrics to the app's `trace_logger` if set and
+        send to `log.debug` if this controller if `self.debugging`.
+
+        Precondition: metrics are parsed and in proper format.
+        """
+        if trans.app.trace_logger:
+            for label, time, kwargs in metrics:
+                trans.app.trace_logger.log(label, event_time=int(time.timestamp()), **kwargs)
+        elif self.debugging:
+            for label, time, kwargs in metrics:
+                log.debug(f'{label} {time} {kwargs}')
```

Comment:
I was wondering whether this interface was meant to handle logs, or metrics or both? I'm particularly interested in adding a Prometheus metrics endpoint to Galaxy, so that metrics can be directly scraped by Prometheus in future. Then there's also sentry for logshipping, and statsd for metrics and I was wondering these different systems are meant to play together? Will a separate interface be introduced for logging?

Reply:
If I recall correctly, this interface was initially pretty narrow and was only meant to be used to capture metrics produced by the client code. Which isn't to say we shouldn't do more with it or that it is even still be used in that capacity currently.

Reply:
Ok thanks for the clarification - that makes sense.